File: fixtures/coding/maintainability-mini-service-go/bulk_import_test.go
package miniservice

import (
	"reflect"
	"testing"
)

func bulkUser() *User {
	return &User{ID: "ops-1", Permissions: []string{"customers:bulk_import"}}
}

func postBulk(app *Application, rows any, user *User) Response {
	if user == nil {
		user = adminUser()
	}
	return app.Handle(Request{Method: "POST", Path: "/customers/bulk-import", JSON: map[string]any{"rows": rows}, User: user})
}

func TestBulkImportAcceptsValidRowsAndReturnsSummary(t *testing.T) {
	app := BuildApp()
	response := postBulk(app, []any{
		map[string]any{"name": "Ada Lovelace", "email": "ada@example.com", "plan": "pro", "tags": []any{"vip", "math"}},
		map[string]any{"name": "Grace Hopper", "email": "GRACE@example.com", "plan": "enterprise"},
	}, nil)

	if response.StatusCode != 200 {
		t.Fatalf("status = %d, want 200", response.StatusCode)
	}
	if response.Body["accepted_count"] != 2 || response.Body["rejected_count"] != 0 {
		t.Fatalf("counts = %#v", response.Body)
	}
	accepted := response.Body["accepted"].([]map[string]any)
	if got := []any{accepted[0]["email"], accepted[1]["email"]}; !reflect.DeepEqual(got, []any{"ada@example.com", "grace@example.com"}) {
		t.Fatalf("accepted emails = %#v", got)
	}
	if !reflect.DeepEqual(response.Body["rejected"], []map[string]any{}) {
		t.Fatalf("rejected = %#v", response.Body["rejected"])
	}
	if got := customerEmails(app.Repository.ListCustomers()); !reflect.DeepEqual(got, []string{"ada@example.com", "grace@example.com"}) {
		t.Fatalf("repository emails = %#v", got)
	}
	if len(app.AuditLog.ListEvents()) != 0 {
		t.Fatalf("unexpected audit events")
	}
}

func TestBulkImportRequiresBulkPermissionAndDoesNotMutateState(t *testing.T) {
	app := BuildApp()
	response := postBulk(app, []any{map[string]any{"name": "Ada", "email": "ada@example.com"}}, &User{ID: "viewer"})
	if response.StatusCode != 403 {
		t.Fatalf("status = %d, want 403", response.StatusCode)
	}
	if !reflect.DeepEqual(response.Body, map[string]any{"error": "forbidden"}) {
		t.Fatalf("body = %#v", response.Body)
	}
	if len(app.Repository.ListCustomers()) != 0 || len(app.AuditLog.ListEvents()) != 0 {
		t.Fatalf("state mutated")
	}
}

func TestBulkImportAcceptsDedicatedPermissionWithoutAdminRole(t *testing.T) {
	app := BuildApp()
	response := postBulk(app, []any{map[string]any{"name": "Ada", "email": "ada@example.com"}}, bulkUser())
	if response.StatusCode != 200 {
		t.Fatalf("status = %d, want 200", response.StatusCode)
	}
	if response.Body["accepted_count"] != 1 {
		t.Fatalf("accepted_count = %#v", response.Body["accepted_count"])
	}
	if app.Repository.FindByEmail("ada@example.com") == nil {
		t.Fatalf("customer not inserted")
	}
}

func TestBulkImportRejectsInvalidRowsWithIndexedErrorsAndAuditEvent(t *testing.T) {
	app := BuildApp()
	response := postBulk(app, []any{
		map[string]any{"name": "", "email": "bad", "plan": "gold"},
		map[string]any{"name": "Valid Customer", "email": "valid@example.com", "plan": "free", "tags": []any{"new"}},
		map[string]any{"name": "Bad Tags", "email": "tags@example.com", "tags": []any{"ok", 99}},
	}, nil)
	if response.StatusCode != 200 {
		t.Fatalf("status = %d, want 200", response.StatusCode)
	}
	if response.Body["accepted_count"] != 1 || response.Body["rejected_count"] != 2 {
		t.Fatalf("counts = %#v", response.Body)
	}
	accepted := response.Body["accepted"].([]map[string]any)
	if accepted[0]["email"] != "valid@example.com" {
		t.Fatalf("accepted = %#v", accepted)
	}
	rejected := response.Body["rejected"].([]map[string]any)
	if rejected[0]["index"] != 0 {
		t.Fatalf("first rejected index = %#v", rejected[0])
	}
	firstErrors := rejected[0]["errors"].([]string)
	assertContains(t, firstErrors, "name is required")
	assertContains(t, firstErrors, "email must be valid")
	assertContains(t, firstErrors, "plan is invalid")
	wantSecond := map[string]any{"index": 2, "email": "tags@example.com", "errors": []string{"tags must be a list of strings"}}
	if !reflect.DeepEqual(rejected[1], wantSecond) {
		t.Fatalf("second rejected = %#v", rejected[1])
	}

	events := app.AuditLog.ListEvents()
	if len(events) != 1 {
		t.Fatalf("events = %#v", events)
	}
	if events[0].Type != "customers.bulk_import.rejected" || events[0].ActorID != "admin-1" {
		t.Fatalf("event = %#v", events[0])
	}
	if !reflect.DeepEqual(events[0].Payload, map[string]any{"accepted_count": 1, "rejected_count": 2}) {
		t.Fatalf("payload = %#v", events[0].Payload)
	}
}

func TestBulkImportRejectsExistingAndInBatchDuplicateEmails(t *testing.T) {
	app := BuildApp()
	app.Handle(Request{Method: "POST", Path: "/customers", JSON: map[string]any{"name": "Existing", "email": "existing@example.com"}, User: adminUser()})
	response := postBulk(app, []any{
		map[string]any{"name": "Existing Again", "email": "existing@example.com"},
		map[string]any{"name": "First", "email": "dupe@example.com"},
		map[string]any{"name": "Second", "email": "DUPE@example.com"},
	}, nil)
	if response.StatusCode != 200 {
		t.Fatalf("status = %d, want 200", response.StatusCode)
	}
	accepted := response.Body["accepted"].([]map[string]any)
	if response.Body["accepted_count"] != 1 || accepted[0]["email"] != "dupe@example.com" {
		t.Fatalf("accepted = %#v", response.Body)
	}
	wantRejected := []map[string]any{
		{"index": 0, "email": "existing@example.com", "errors": []string{"customer already exists"}},
		{"index": 2, "email": "dupe@example.com", "errors": []string{"duplicate email in import"}},
	}
	if !reflect.DeepEqual(response.Body["rejected"], wantRejected) {
		t.Fatalf("rejected = %#v", response.Body["rejected"])
	}
	if got := customerEmails(app.Repository.ListCustomers()); !reflect.DeepEqual(got, []string{"existing@example.com", "dupe@example.com"}) {
		t.Fatalf("repository emails = %#v", got)
	}
}

func TestBulkImportRejectsMissingOrNonListRowsPayload(t *testing.T) {
	app := BuildApp()
	missing := app.Handle(Request{Method: "POST", Path: "/customers/bulk-import", JSON: map[string]any{}, User: adminUser()})
	if missing.StatusCode != 400 || !reflect.DeepEqual(missing.Body, map[string]any{"error": "rows must be a list"}) {
		t.Fatalf("missing = %#v", missing)
	}
	wrongType := app.Handle(Request{Method: "POST", Path: "/customers/bulk-import", JSON: map[string]any{"rows": "not-list"}, User: adminUser()})
	if wrongType.StatusCode != 400 || !reflect.DeepEqual(wrongType.Body, map[string]any{"error": "rows must be a list"}) {
		t.Fatalf("wrong type = %#v", wrongType)
	}
}

func TestBulkImportPreservesExistingListRouteAfterImport(t *testing.T) {
	app := BuildApp()
	postBulk(app, []any{map[string]any{"name": "Ada", "email": "ada@example.com"}, map[string]any{"name": "Grace", "email": "grace@example.com"}}, nil)
	listed := app.Handle(Request{Method: "GET", Path: "/customers", User: adminUser()})
	if listed.StatusCode != 200 {
		t.Fatalf("status = %d, want 200", listed.StatusCode)
	}
	customers := listed.Body["customers"].([]map[string]any)
	if len(customers) != 2 {
		t.Fatalf("listed customers = %#v", customers)
	}
	if got := []any{customers[0]["email"], customers[1]["email"]}; !reflect.DeepEqual(got, []any{"ada@example.com", "grace@example.com"}) {
		t.Fatalf("listed = %#v", got)
	}
}

func assertContains(t *testing.T, values []string, want string) {
	t.Helper()
	for _, value := range values {
		if value == want {
			return
		}
	}
	t.Fatalf("%q not found in %#v", want, values)
}

func customerEmails(customers []Customer) []string {
	emails := make([]string, 0, len(customers))
	for _, customer := range customers {
		emails = append(emails, customer.Email)
	}
	return emails
}

File: fixtures/coding/maintainability-mini-service-ts/tests/bulk-import.test.ts
import { describe, expect, test } from "vitest";
import { buildApp } from "../src/container";
import type { Application } from "../src/router";
import type { Request, User } from "../src/models";

const admin = (): User => ({ id: "admin-1", role: "admin" });
const bulkUser = (): User => ({ id: "ops-1", permissions: ["customers:bulk_import"] });

function postBulk(app: Application, rows: unknown, user: User = admin()) {
  return app.handle({ method: "POST", path: "/customers/bulk-import", json: { rows }, user });
}

describe("bulk customer import", () => {
  test("accepts valid rows and returns summary", () => {
    const app = buildApp();
    const response = postBulk(app, [
      { name: "Ada Lovelace", email: "ada@example.com", plan: "pro", tags: ["vip", "math"] },
      { name: "Grace Hopper", email: "GRACE@example.com", plan: "enterprise" },
    ]);

    expect(response.statusCode).toBe(200);
    expect(response.body.accepted_count).toBe(2);
    expect(response.body.rejected_count).toBe(0);
    expect((response.body.accepted as Array<{ email: string }>).map((item) => item.email)).toEqual([
      "ada@example.com",
      "grace@example.com",
    ]);
    expect(response.body.rejected).toEqual([]);
    expect(app.repository.listCustomers().map((customer) => customer.email)).toEqual(["ada@example.com", "grace@example.com"]);
    expect(app.auditLog.listEvents()).toEqual([]);
  });

  test("requires bulk permission and does not mutate state", () => {
    const app = buildApp();
    const response = postBulk(app, [{ name: "Ada", email: "ada@example.com" }], { id: "viewer" });

    expect(response.statusCode).toBe(403);
    expect(response.body).toEqual({ error: "forbidden" });
    expect(app.repository.listCustomers()).toEqual([]);
    expect(app.auditLog.listEvents()).toEqual([]);
  });

  test("accepts dedicated permission without admin role", () => {
    const app = buildApp();
    const response = postBulk(app, [{ name: "Ada", email: "ada@example.com" }], bulkUser());

    expect(response.statusCode).toBe(200);
    expect(response.body.accepted_count).toBe(1);
    expect(app.repository.findByEmail("ada@example.com")).toBeDefined();
  });

  test("rejects invalid rows with indexed errors and audit event", () => {
    const app = buildApp();
    const response = postBulk(app, [
      { name: "", email: "bad", plan: "gold" },
      { name: "Valid Customer", email: "valid@example.com", plan: "free", tags: ["new"] },
      { name: "Bad Tags", email: "tags@example.com", tags: ["ok", 99] },
    ]);

    expect(response.statusCode).toBe(200);
    expect(response.body.accepted_count).toBe(1);
    expect(response.body.rejected_count).toBe(2);
    expect((response.body.accepted as Array<{ email: string }>)[0]?.email).toBe("valid@example.com");

    const rejected = response.body.rejected as Array<{ index: number; email: string; errors: string[] }>;
    expect(rejected[0]?.index).toBe(0);
    expect(rejected[0]?.errors).toContain("name is required");
    expect(rejected[0]?.errors).toContain("email must be valid");
    expect(rejected[0]?.errors).toContain("plan is invalid");
    expect(rejected[1]).toEqual({ index: 2, email: "tags@example.com", errors: ["tags must be a list of strings"] });

    const events = app.auditLog.listEvents();
    expect(events).toHaveLength(1);
    expect(events[0]).toEqual({
      type: "customers.bulk_import.rejected",
      actorId: "admin-1",
      payload: { accepted_count: 1, rejected_count: 2 },
    });
  });

  test("rejects existing and in-batch duplicate emails", () => {
    const app = buildApp();
    app.handle({ method: "POST", path: "/customers", json: { name: "Existing", email: "existing@example.com" }, user: admin() });

    const response = postBulk(app, [
      { name: "Existing Again", email: "existing@example.com" },
      { name: "First", email: "dupe@example.com" },
      { name: "Second", email: "DUPE@example.com" },
    ]);

    expect(response.statusCode).toBe(200);
    expect(response.body.accepted_count).toBe(1);
    expect((response.body.accepted as Array<{ email: string }>)[0]?.email).toBe("dupe@example.com");
    expect(response.body.rejected).toEqual([
      { index: 0, email: "existing@example.com", errors: ["customer already exists"] },
      { index: 2, email: "dupe@example.com", errors: ["duplicate email in import"] },
    ]);
    expect(app.repository.listCustomers().map((customer) => customer.email)).toEqual(["existing@example.com", "dupe@example.com"]);
  });

  test("rejects missing or non-list rows payload", () => {
    const app = buildApp();

    const missing = app.handle({ method: "POST", path: "/customers/bulk-import", json: {}, user: admin() });
    expect(missing.statusCode).toBe(400);
    expect(missing.body).toEqual({ error: "rows must be a list" });

    const wrongType = app.handle({ method: "POST", path: "/customers/bulk-import", json: { rows: "not-list" }, user: admin() });
    expect(wrongType.statusCode).toBe(400);
    expect(wrongType.body).toEqual({ error: "rows must be a list" });
  });

  test("preserves existing list route after import", () => {
    const app = buildApp();
    postBulk(app, [{ name: "Ada", email: "ada@example.com" }, { name: "Grace", email: "grace@example.com" }]);

    const listed = app.handle({ method: "GET", path: "/customers", user: admin() });
    expect(listed.statusCode).toBe(200);
    expect((listed.body.customers as Array<{ email: string }>).map((customer) => customer.email)).toEqual([
      "ada@example.com",
      "grace@example.com",
    ]);
  });
});
